rasterio/_pyvsi.pyx

```diff
@@ -0,0 +1,369 @@
+
+
+cdef int pyvsi_truncate(void *pFile, vsi_l_offset nNewSize) with gil:
+    print("truncate")
+
+
+cdef int pyvsi_close(void *pFile) except -1 with gil:
+    # Optional
+    print("close")
+    cdef object file_wrapper = <object>pFile
+    del _FILESYSTEM_INFO[file_wrapper._pyvsi_path]
+    return 0
+
+
+cdef int install_rasterio_pyvsi_plugin(VSIFilesystemPluginCallbacksStruct *callbacks_struct):
+    """Install handlers for python file-like objects if it isn't already installed."""
+    cdef int install_status
+    if VSIFileManager.GetHandler("") == VSIFileManager.GetHandler(FILESYSTEM_PREFIX_BYTES):
+        log.debug("Installing PyVSI filesystem handler plugin...")
+        install_status = VSIInstallPluginHandler(FILESYSTEM_PREFIX_BYTES, callbacks_struct)
+        return install_status
+    return 0
+
+
+cdef class PyVSIFileBase:
+
+    cdef VSIFilesystemPluginCallbacksStruct* _vsif
+
+    def __cinit__(self, file_or_bytes, *args, **kwargs):
+        self._vsif = VSIAllocFilesystemPluginCallbacksStruct()
+        # pUserData will be set later
+        self._vsif.open = <VSIFilesystemPluginOpenCallback>pyvsi_open
+
+        self._vsif.tell = <VSIFilesystemPluginTellCallback>pyvsi_tell
+        self._vsif.seek = <VSIFilesystemPluginSeekCallback>pyvsi_seek
+        self._vsif.read = <VSIFilesystemPluginReadCallback>pyvsi_read
+        # self._vsif.eof = <VSIFilesystemPluginEofCallback>pyvsi_eof
+        self._vsif.close = <VSIFilesystemPluginCloseCallback>pyvsi_close
+
+    def __dealloc__(self):
+        if self._vsif is not NULL:
+            self._vsif.pUserData = NULL
+            VSIFreeFilesystemPluginCallbacksStruct(self._vsif)
+            self._vsif = NULL
+
+    def __init__(self, file_or_bytes=None, dirname=None, filename=None, ext='.tif'):
+        """A file in an in-memory filesystem.
+
+        Parameters
+        ----------
+        file_or_bytes : file or bytes
+            A file opened in binary mode or bytes
+        filename : str
+            A filename for the in-memory file under /vsimem
+        ext : str
+            A file extension for the in-memory file under /vsimem. Ignored if
+            filename was provided.
+
+        """
+        if isinstance(file_or_bytes, (bytes, str)) or not hasattr(file_or_bytes, "read"):
+            raise TypeError("PyVSIFile expects file-like objects only.")
+
+        # Make an in-memory directory specific to this dataset to help organize
+        # auxiliary files.
+        # XXX: Do we really need directories?
+        self._dirname = dirname or str(uuid4())
+
+        if filename:
+            # GDAL's SRTMHGT driver requires the filename to be "correct" (match
+            # the bounds being written)
+            self.name = "{0}{1}/{2}".format(FILESYSTEM_PREFIX, self._dirname, filename)
+        else:
+            # GDAL 2.1 requires a .zip extension for zipped files.
+            self.name = "{0}{1}/{1}.{2}".format(FILESYSTEM_PREFIX, self._dirname, ext.lstrip('.'))
```

Comment:
@sgillies Any idea if GDAL tries to use filename extensions to determine things? Otherwise I could just remove the `ext` keyword. It isn't provided by the rasterio `open` function so I don't think any public APIs will be providing it.

Reply:
@djhoese let's remove it.

In https://github.com/mapbox/rasterio/issues/2064 I proposed that rasterio releases support GDAL minor releases made in the prior 24 months from the anticipated rasterio release date. If rasterio 1.3.0 were to come out in November 2021, that might mean supporting GDAL 2.4 (2.4.4. released in January 2020). The GDAL project isn't maintaining 2.4 anymore, however, and there will never be a GDAL 2.4.5, so perhaps that's an argument in favor of dropping support for it.

Reply:
I've removed the `ext` keyword argument for the PythonVSIFile interface.